File: packages/Demo/scripts/node/pca.js
//name: PCA Node
//description: Principal Component Analysis
//help-url: https://en.wikipedia.org/wiki/Principal_component_analysis
//language: nodejs
//tags: demo, hide-suggestions
//sample: cars.csv
//input: dataframe table {columns:numerical} [Input data table]
//input: column_list features {type:numerical;allowNulls:false} [Features columns]
//input: int components = 2 [Number of components]
//input: bool center = TRUE [Indicating whether the variables should be shifted to be zero centered]
//input: bool scale = TRUE [Indicating whether the variables should be scaled to have unit variance before the analysis takes place]
//output: dataframe result {action:join(table)} [PCA components]

const PCA = require('ml-pca');
const array = table.toArray();
const pca = new PCA(array, {'scale': scale, 'center': center});
const predicted = pca.predict(array, {'nComponents': components});
let result = new DataFrame(predicted, Array.from({length: components}, (v, i) => 'PCA' + i.toString()));

File: packages/Samples/scripts/r/pca.R
#name: PCA R
#description: Principal Component Analysis
#reference: https://en.wikipedia.org/wiki/Principal_component_analysis
#language: r
#tags: demo
#sample: cars.csv
#input: dataframe T {columns:numerical} [Input data]
#input: column_list columns {type:numerical;allowNulls:false} [Input data columns]
#input: int numComp = 2 [Number of components]
#input: bool center = TRUE [Indicating whether the variables should be shifted to be zero centered]
#input: bool scale = TRUE [Indicating whether the variables should be scaled to have unit variance before the analysis takes place]
#output: dataframe result {action:join(T)} [PCA components]

require(stats)

# Preform PCA and select required number of components
T.pca <- prcomp(T, center=center, scale.=scale)
result <- predict(T.pca, T)
result = result[,1:numComp]
